web_domain_field/static/src/js/pyeval.js

```diff
@@ -0,0 +1,233 @@
+        var val = context[k];
+
+        if (val === null) { continue; }
+        if (val.constructor === Array) {
+            context[k] = wrapping_list.fromJSON(val);
+        } else if (val.constructor === Object
+                   && !py.PY_isInstance(val, py.object)) {
+            context[k] = wrapping_dict.fromJSON(val);
+        }
+    }
+    return context;
+}
+
+function ensure_evaluated (args, kwargs) {
+    for (var i=0; i<args.length; ++i) {
+        args[i] = eval_arg(args[i]);
+    }
+    for (var k in kwargs) {
+        if (!kwargs.hasOwnProperty(k)) { continue; }
+        kwargs[k] = eval_arg(kwargs[k]);
+    }
+}
+function eval_contexts (contexts, evaluation_context) {
+    evaluation_context = _.extend(pyeval.context(), evaluation_context || {});
+    return _(contexts).reduce(function (result_context, ctx) {
+        // __eval_context evaluations can lead to some of `contexts`'s
+        // values being null, skip them as well as empty contexts
+        if (_.isEmpty(ctx)) { return result_context; }
+        if (_.isString(ctx)) {
+            // wrap raw strings in context
+            ctx = { __ref: 'context', __debug: ctx };
+        }
+        var evaluated = ctx;
+        switch(ctx.__ref) {
+        case 'context':
+            evaluation_context.context = evaluation_context;
+            evaluated = py.eval(ctx.__debug, wrap_context(evaluation_context));
+            break;
+        case 'compound_context':
+            var eval_context = eval_contexts([ctx.__eval_context]);
+            evaluated = eval_contexts(
+                ctx.__contexts, _.extend({}, evaluation_context, eval_context));
+            break;
+        }
+        // add newly evaluated context to evaluation context for following
+        // siblings
+        _.extend(evaluation_context, evaluated);
+        return _.extend(result_context, evaluated);
+    }, {});
+}
+
+/** end of unmodified methods copied from pyeval **/
+
+// We need to override the original method to be able to call our
+//specialized version of pyeval for domain fields
+function eval_arg (arg) {
+    if (typeof arg !== 'object' || !arg.__ref) { return arg; }
+    switch(arg.__ref) {
+    case 'domain': case 'compound_domain':
+        return domain_field_pyeval('domains', [arg]);
+    case 'context': case 'compound_context':
+        return original_pyeval('contexts', [arg]);
+    default:
+        throw new Error(_t("Unknown nonliteral type ") + ' ' + arg.__ref);
+    }
+}
+
+// override eval_domains to add 3 lines in order to be able to use a field
+//value as domain
+function eval_domains (domains, evaluation_context) {
+    evaluation_context = _.extend(pyeval.context(), evaluation_context ||
+    {});
+    var result_domain = [];
+    _(domains).each(function (domain) {
+        if (_.isString(domain)) {
+           // Modified part or the original method
+           if(domain in evaluation_context) {
+                result_domain.push.apply(
```

Comment:
Hi @rousseldenis, @lmignon ,
I tried this port of v9 module but I get an error on this apply function here.
Don't know what could I do to help, so here is at least the error I get and my example code.

```
TypeError: second argument to Function.prototype.apply must be an array

http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4267
Traceback:
eval_domains/<@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4267:230
_.forEach@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:12:558
_.mixin/</_.prototype[name]@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:69:521
eval_domains@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4267:141
eval_domains/<@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4269:301
_.forEach@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:12:558
_.mixin/</_.prototype[name]@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:69:521
eval_domains@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4267:141
eval_domains/<@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4269:301
_.forEach@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:12:558
_.mixin/</_.prototype[name]@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:69:521
eval_domains@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4267:141
domain_field_pyeval@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4271:24
eval_arg@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4266:61
ensure_evaluated@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:4260:11
call@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:1476:1
name_search@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:1758:1038
get_search_result@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:2224:25
source@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:2332:103
_search@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:1461:142
$.widget/</proxiedPrototype[prop]</<@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:1021:328
search@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:1461:8
$.widget/</proxiedPrototype[prop]</<@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:1021:328
$.widget.bridge/$.fn[name]/<@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:1026:13
each@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:370:758
each@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:358:636
$.widget.bridge/$.fn[name]@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:1023:248
render_editable/<@http://localhost:8069/web/content/307-c8e6145/web.assets_backend.js:2327:1041
dispatch@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:637:378
add/elemData.handle@http://localhost:8069/web/content/293-31cc17f/web.assets_common.js:610:151
```

My code looks like this :
```
original_order_products_domain = fields.Char(
        'product.product', compute="_compute_original_order_products_domain")
    original_order_id = fields.Many2one(
        'sale.order',
        string='Original sale order'
    )

    @api.multi
    @api.depends('original_order_id')
    def _compute_original_order_products_domain(self):
        for r in self:
            order_lines = self.env['sale.order.line'].search(
                [('order_id', '=', r.original_order_id.id)])
            if order_lines:
                product_ids = order_lines.mapped('product_id').ids
                r.original_order_products_domain = json.dumps(
                    "[('id', 'in', %s)]" % product_ids)
```
And on the view :
```
<field name="original_order_products_domain" readonly="1"/>
<field name="product_id" domain="original_order_products_domain"/>
```